### PR TITLE
Library scene

### PR DESCRIPTION
**Pull request by vandermesis**

### CODE REVIEW

File: Templates/Clean Swift/Scene.xctemplate/UITableViewControllerXIB/___FILEBASENAME___Controller.swift [hunk 1]
```diff
@@ -45,6 +45,7 @@ private extension ___VARIABLE_sceneName___Controller {
     private func setupTableView() {
         //TODO: Register custom cell
 //        tableView.register(cellType: ___VARIABLE_sceneName___TableViewCell.self)
+        tableView.reloadData()
```
Comment (SredzinskiPatryk) [inline Templates/Clean Swift/Scene.xctemplate/UITableViewControllerXIB/___FILEBASENAME___Controller.swift]:
```suggestion
```

you don't need this, as setupTableView is called in viewDidLoad

File: Templates/Clean Swift/UnitTestsXC.xctemplate/___FILEBASENAME___WorkerTests.swift [hunk 2]
```diff
@@ -31,4 +37,52 @@ final class ___VARIABLE_sceneName___WorkerTests: XCTestCase {
...
+                self.receivedError = error
+            }
+        }
+        // When
+
+        // Then
+
+    }
+
+    func testWorkerIsFetchingPlacesIsSuccess() {
```
Comment (SredzinskiPatryk) [inline Templates/Clean Swift/UnitTestsXC.xctemplate/___FILEBASENAME___WorkerTests.swift]:
```suggestion
    func testWorkerIsFetchingIsSuccess() {
```

File: Templates/Clean Swift/UnitTestsXC.xctemplate/___FILEBASENAME___WorkerTests.swift [hunk 3]
```diff
@@ -31,4 +37,52 @@ final class ___VARIABLE_sceneName___WorkerTests: XCTestCase {
...
+                self.receivedError = error
+            }
+        }
+        // When
+
+        // Then
+
+    }
+
+    func testWorkerIsFetchingPlacesIsFailure() {
```
Comment (SredzinskiPatryk) [inline Templates/Clean Swift/UnitTestsXC.xctemplate/___FILEBASENAME___WorkerTests.swift]:
```suggestion
    func testWorkerIsFetchingIsFailure() {
```

File: ZeroWasteFriendly/Repository/LibraryRepository.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            case .failure(let error):
+                completion?(.failure(error))
+            }
+        }
+    }
+}
+
+private extension LibraryRepositoryImpl {
+
+    private func convertRecordsToPosts(records: [CKRecord]) -> [Post] {
```
Comment (SredzinskiPatryk) [inline ZeroWasteFriendly/Repository/LibraryRepository.swift:36]: is it possible to use Codable features to allow automatic conversion?

File: ZeroWasteFriendly/Scenes/Library/LibraryController.swift [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        setupTableView()
+        interactor.getPosts()
+    }
+}
+
+private extension LibraryController {
+
+    private func setupTableView() {
+        tableView.register(cellType: LibraryTableViewCell.self)
+        tableView.separatorStyle = .none
```
Comment (SredzinskiPatryk) [inline ZeroWasteFriendly/Scenes/Library/LibraryController.swift]: move this separator style to XIB
I'd do it in template too, as you mostly don't want separators

File: ZeroWasteFriendly/Scenes/Library/LibraryController.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        return cell
+    }
+}
+
+extension LibraryController: UITableViewDelegate {
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let selectedPostID = postsDataSource[indexPath.row].id
+        interactor.didSelectPost(id: selectedPostID)
+        tableView.reloadData()
```
Comment (SredzinskiPatryk) [inline ZeroWasteFriendly/Scenes/Library/LibraryController.swift]: reloadData should be called by presenter
didSelectPost

File: ZeroWasteFriendly/Scenes/Library/LibraryPresenter.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    func expandPost(id: String) {
+        var convertedPosts = [PostDisplayable]()
+        postsDisplayable.forEach { post in
+            if post.id == id {
+                post.expanded.toggle()
+            }
+            convertedPosts.append(post)
+        }
+        controller?.displayPosts(posts: convertedPosts)
```
Comment (SredzinskiPatryk) [inline ZeroWasteFriendly/Scenes/Library/LibraryPresenter.swift:37]: you're calling controller, and it reloads data twice

File: ZeroWasteFriendly/Scenes/Library/LibraryRouter.swift [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+final class LibraryRouterImpl {
+
+    weak var controller: MainViewController?
+}
+
+extension LibraryRouterImpl: LibraryRouter {
+
+    func navigateTo() {
+        let InfoController = LibraryCreator().getController()
```
Comment (SredzinskiPatryk) [inline ZeroWasteFriendly/Scenes/Library/LibraryRouter.swift]: mark it with TODO
update template

File: ZeroWasteFriendly/Scenes/Library/LibraryTableViewCell/LibraryTableViewCell.swift [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    @IBOutlet private weak var shadowView: UIView!
+
+    func setup(with post: PostDisplayable) {
+        titleLabel.text = post.title
+        contentLabel.text = post.content
+        addCellShadow()
+    }
+}
+
+private extension LibraryTableViewCell {
```
Comment (SredzinskiPatryk) [inline ZeroWasteFriendly/Scenes/Library/LibraryTableViewCell/LibraryTableViewCell.swift:28]: this could be a general extension to uiview

**Review by SredzinskiPatryk — CHANGES_REQUESTED** [review]

File: Templates/Clean Swift/Scene.xctemplate/UITableViewControllerXIB/___FILEBASENAME___Controller.swift (hunk 1 above)
Comment (vandermesis) [reply Templates/Clean Swift/Scene.xctemplate/UITableViewControllerXIB/___FILEBASENAME___Controller.swift]: Sorry that was mistake.

Comment (vandermesis) [reply Templates/Clean Swift/Scene.xctemplate/UITableViewControllerXIB/___FILEBASENAME___Controller.swift]: I was going to put that line into display... method :/

**Review by SredzinskiPatryk — APPROVED** [review]

**Pull request merged**